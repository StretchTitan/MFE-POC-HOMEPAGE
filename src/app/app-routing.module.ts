import { NgModule } from '@angular/core';
import { Routes, RouterModule, UrlSegment } from '@angular/router';
import { EquipmentComponent } from './equipment/equipment.component';
import { ChannelsComponent } from './channels/channels.component';
import { EmptyComponent } from './empty/empty.component';

const routes: Routes = [
  {
    path: 'home',
    component: EmptyComponent,
  },
  {
    path: 'home/equipment',
    component: EquipmentComponent,
    pathMatch: 'full',
  },
  {
    path: 'home/channels',
    component: ChannelsComponent,
    pathMatch: 'full',
  },
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule],
})
export class AppRoutingModule {}
