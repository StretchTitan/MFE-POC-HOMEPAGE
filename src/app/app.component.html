<app-homepage-alert class="homepage-alert-overlay" [ngClass]="{'visible': showInternalAlert}" (toggleInternalAlert)="toggleInternalAlert()"></app-homepage-alert>
<div class="homepage-app">
    <h4>This is the <span style="color: darkgreen">MFE-POC-HOMEPAGE</span> app</h4>
    <p>Welcome to the home page!</p>
    <p *ngIf="(nameState$ | push) as n">Name: {{n.firstName}} {{n.lastName}}</p>
    <p *ngIf="(storeNameState$ | push) as sn">Store Name: {{sn.firstName}} {{sn.lastName}}</p>
    <button (click)="clearName()">Clear Name</button>
    <ng-template #error>Loading failed...</ng-template>
    <ng-template #loading>Loading...</ng-template>
    <mfe-widget *axLazyElement="'http://local.spectrum-poc.net:4203/shared-widget.js'; loadingTemplate: loading; errorTemplate: error" [counter$]="counter$"></mfe-widget>
    <mfe-card *axLazyElement="'http://local.spectrum-poc.net:4203/shared-card.js'; loadingTemplate: loading; errorTemplate: error"></mfe-card>
    <button (click)="flipToggle()">Flip Toggle</button>
    <p style="font-weight: bold;" [ngStyle]="{'color': (toggleState$ | async) ? 'darkgreen' : 'red'}">Toggle: {{ (toggleState$ | async) ? 'ON' : 'OFF' }}</p>
    <mfe-homepage-widget *axLazyElement="'http://local.spectrum-poc.net:4204/shared-homepage-widget.js'; loadingTemplate: loading; errorTemplate: error"></mfe-homepage-widget>
    <button class="alert-button" (click)="showAlert()">Show Alert</button>
    <button class="panel-button" (click)="toggleInternalAlert()">Show Internal Alert</button>
    <button class="panel-button" (click)="showPanel()">Show Panel</button>
    <hr/>
    <button (click)="getToken()">Get Token</button> <button (click)="getApi()">Make API Call</button> <button *ngIf="(display$ | async)" (click)="clearResult()">Clear Result</button>
    <div *ngIf="(display$ | async) as d">
        <h4>Result</h4>
        <p>{{d | json}}</p>
    </div>
</div> 
<div class="nav">
    <a class="nav--link" routerLink="/home/equipment">Equipment</a>
    <a class="nav--link" routerLink="/home/channels">Channels</a>
</div>
<router-outlet></router-outlet>
