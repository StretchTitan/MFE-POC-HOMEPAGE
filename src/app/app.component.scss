:host {
    display: flex;
    flex: 1;
}

.homepage-app {
    padding: 18px;
    margin: 4px;
    border: 2px dashed darkgreen;
    width: 100%;
}