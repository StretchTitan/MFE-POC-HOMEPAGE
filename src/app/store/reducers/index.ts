import {
  ActionReducerMap,
  MetaReducer
} from '@ngrx/store';
import { environment } from '../../../environments/environment';
import { toggleFeatureKey, toggleReducer, ToggleState } from './toggle/toggle.reducer';
import { WrapperState, wrapperFeatureKey, wrapperReducer } from './wrapper/wrapper.reducer';

export interface AppState {
  [toggleFeatureKey]: ToggleState;
  [wrapperFeatureKey]: WrapperState;
}

export const reducers: ActionReducerMap<AppState> = {
  [toggleFeatureKey]: toggleReducer,
  [wrapperFeatureKey]: wrapperReducer,
};

export const metaReducers: MetaReducer<AppState>[] = !environment.production ? [] : [];
