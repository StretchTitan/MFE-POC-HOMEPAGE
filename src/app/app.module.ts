import { BrowserModule } from '@angular/platform-browser';
import { NgModule, Injector, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { createCustomElement } from '@angular/elements';
import { LazyElementsModule, LAZY_ELEMENTS_REGISTRY } from '@angular-extensions/elements';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';
import { EquipmentComponent } from './equipment/equipment.component';
import { ChannelsComponent } from './channels/channels.component';
import { PushPipe } from './push.pipe';
import { CustomLazyRegistry } from './custom-lazy-registry';
import { EmptyComponent } from './empty/empty.component';

@NgModule({
  declarations: [
    AppComponent,
    EquipmentComponent,
    ChannelsComponent,
    PushPipe,
    EmptyComponent,
  ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    HttpClientModule,
    LazyElementsModule,
  ],
  schemas: [CUSTOM_ELEMENTS_SCHEMA],
  providers: [
    PushPipe,
    {
      provide: LAZY_ELEMENTS_REGISTRY,
      useClass: CustomLazyRegistry
    }
  ],
  // Uncomment this line and run ng serve to only run this app locally
  // bootstrap: [AppComponent],
  entryComponents: [AppComponent],
})

export class AppModule {
  constructor(private injector: Injector) {
    const appElement = createCustomElement(AppComponent, { injector: this.injector });
    customElements.define('mfe-poc-homepage', appElement);
  }

  ngDoBootstrap() { }
}
