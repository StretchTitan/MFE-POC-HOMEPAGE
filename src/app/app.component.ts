import { Component, Input, Output, EventEmitter, SimpleChanges, OnChanges, ViewChild, OnDestroy } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router, RouterOutlet } from '@angular/router';
import { interval, Observable, Subject } from 'rxjs';
import { map, take, takeUntil } from 'rxjs/operators';
import { Store } from '@ngrx/store';

import { selectToggle } from './store/selectors/toggle/toggle.selectors';
import { toggle } from './store/actions/toggle/toggle.actions';
import { hydrateWrapper } from './store/actions/wrapper/wrapper.actions';
import { selectWrapperNameState } from './store/selectors/wrapper/wrapper.selectors';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
})
export class AppComponent implements OnChanges, OnDestroy {
  @Input() nameState$: Observable<{ firstName: string, lastName: string }>;
  @Input() routerState: string;
  @Output() msg = new EventEmitter<any>();
  @ViewChild(RouterOutlet) outlet: RouterOutlet;
  toggleState$: Observable<boolean>;
  destroySubs$ = new Subject();
  storeNameState$: Observable<{ firstName: string, lastName: string }>;
  display: any;
  count = 1;
  counter$ = interval(1000).pipe(take(6), map(() => this.count++));

  constructor(private http: HttpClient, private router: Router, private store: Store) {
    this.toggleState$ = store.select(selectToggle);
    this.storeNameState$ = store.select(selectWrapperNameState);
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (changes.routerState && changes.routerState.firstChange) {
      this.router.navigate([changes.routerState.currentValue]);
    }

    if (changes.nameState$ && changes.nameState$.firstChange) {
      this.nameState$.pipe(
        takeUntil(this.destroySubs$)
      ).subscribe(name => this.store.dispatch(hydrateWrapper(name)));
    }
  }

  ngOnDestroy(): void {
    this.destroySubs$.next();
    this.destroySubs$.complete();
  }

  clearName() {
    this.msg.emit({ action: 'clearName' });
  }

  flipToggle() {
    this.store.dispatch(toggle());
  }

  getToken() {
    this.http.get('http://local.spectrum-poc.net:4299/token', { withCredentials: true })
      .subscribe(
        data => this.display = data,
        error => this.display = `${error.message}`,
      );
  }

  getApi() {
    this.http.get('http://local.spectrum-poc.net:4299/api', { withCredentials: true })
      .subscribe(
        data => this.display = data,
        error => this.display = `${error.message}`,
      );
  }

  clearResult() {
    this.display = '';
  }
}
