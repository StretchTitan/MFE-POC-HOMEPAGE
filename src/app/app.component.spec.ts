import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';
import { LazyElementTestingDirective } from '@angular-extensions/elements/testing';
import { marbles } from 'rxjs-marbles/jest';
import { AppComponent } from './app.component';
import { PushPipe } from './push.pipe';
import { CUSTOM_ELEMENTS_SCHEMA, SimpleChange } from '@angular/core';
import { Router } from '@angular/router';
import { provideMockStore } from '@ngrx/store/testing';

describe('AppComponent', () => {
  let component: AppComponent;
  let fixture: ComponentFixture<AppComponent>;
  let router: Router;
  let httpTestingController: HttpTestingController;
  const initialState = {
    wrapper: {
      name: {
        firstName: '',
        lastName: ''
      }
    },
    toggle: {
      toggle: false
    }
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      imports: [RouterTestingModule, HttpClientTestingModule],
      declarations: [AppComponent, PushPipe, LazyElementTestingDirective],
      providers: [
        provideMockStore({ initialState }),
        PushPipe
      ],
    }).compileComponents();

    router = TestBed.inject(Router);
    httpTestingController = TestBed.inject(HttpTestingController);
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AppComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create the app', () => {
    expect(fixture).toMatchSnapshot();
  });

  it(
    'ngOnChanges sets up routerState$ listener',
    marbles((m) => {
      spyOn(router, 'navigate');
      component.routerState = '/home/settings';
      component.ngOnChanges({
        routerState: new SimpleChange(null, component.routerState, true),
      });
      m.flush();
      fixture.detectChanges();
      expect(router.navigate).toHaveBeenCalledTimes(1);
      expect(router.navigate).toHaveBeenCalledWith(['/home/settings']);
    })
  );

  it('ngOnDestroy emits and completes subject to destroy subs', () => {
    spyOn(component.destroySubs$, 'next');
    spyOn(component.destroySubs$, 'complete');
    component.ngOnDestroy();
    fixture.detectChanges();
    expect(component.destroySubs$.next).toHaveBeenCalledTimes(1);
    expect(component.destroySubs$.complete).toHaveBeenCalledTimes(1);
  });

  it('clearName emits msg with proper action', (done) => {
    component.msg.subscribe((value) => {
      expect(value).toEqual({ action: 'clearName' });
      done();
    });
    component.clearName();
  });

  it('getToken makes api call to get token', () => {
    component.getToken();

    const req = httpTestingController.expectOne(
      'http://local.spectrum-poc.net:4299/token'
    );

    req.flush({ token: 'testing' });
    expect(component.display).toEqual({ token: 'testing' });
  });

  it('getApi makes api call to get token', () => {
    component.getApi();

    const req = httpTestingController.expectOne(
      'http://local.spectrum-poc.net:4299/api'
    );

    req.flush({ api: 'testing' });
    expect(component.display).toEqual({ api: 'testing' });
  });

  it('clearDisplay clears display prop', () => {
    component.display = { api: 'testing' };
    component.clearResult();
    expect(component.display).toEqual('');
  });
});
