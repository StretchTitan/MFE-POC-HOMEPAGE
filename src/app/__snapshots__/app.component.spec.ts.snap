// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should create the app 1`] = `
<app-root
  __ngContext__={[Function LRootView]}
  count={[Function Number]}
  counter$={[Function Observable]}
  destroySubs$={[Function Subject]}
  http={[Function HttpClient]}
  msg={[Function EventEmitter]}
  outlet={[Function RouterOutlet]}
  router={[Function Router]}
  store={[Function MockStore]}
  storeNameState$={[Function Store]}
  toggleState$={[Function Store]}
>
  <div
    class="homepage-app"
  >
    <h4>
      This is the 
      <span
        style="color: darkgreen;"
      >
        MFE-POC-HOMEPAGE
      </span>
       app
    </h4>
    <p>
      Welcome to the home page!
    </p>
    
    <p>
      Store Name:  
    </p>
    
    <button>
      Clear Name
    </button>
    
    
    <mfe-widget />
    
    <mfe-card />
    
    <button>
      Flip Toggle
    </button>
    <p>
      Toggle: OFF
    </p>
    <hr />
    <button>
      Get Token
    </button>
    <button>
      Make API Call
    </button>
    
    
  </div><div
    class="nav"
  >
    <a
      class="nav--link"
      href="/home/equipment"
      routerlink="/home/equipment"
    >
      Equipment
    </a>
    <a
      class="nav--link"
      href="/home/channels"
      routerlink="/home/channels"
    >
      Channels
    </a>
  </div><router-outlet />
</app-root>
`;
